accepted/system-range/system-range.md

```diff
@@ -0,0 +1,310 @@
+
+Ranges provide a concise way to provide an `index` and a `length`. This allows
+for making common operations more compact and less confusing at the call side.
+For example, consider this version of the indexer for `String`:
+
+```csharp
+partial class String
+{
+    public string this[Range range] => Substring(range.Index, range.Length);
+}
+```
+
+This allows creating substrings simply by using the indexer:
+
+```csharp
+var helloWorld = "Hello, World!";
+var hello = helloWorld[..5];
+var world = helloWorld[7.:5];
+```
+
+While this might appear to just be minor syntactic sugar it allows for creating
+powerful APIs, such as the indexer for `Tensor<T>` (which you can think of as
+multidimensional array where the dimensions aren't statically known).
+
+```csharp
+partial class Tensor<T>
+{
+    public Tensor<T> this[params Range[] ranges] { get; }
+}
+```
+
+The indexer takes a list of ranges that can be used to get slices of the tensor.
+Imagine you have a two dimensional tensor (in other words a table). Imagine it
+represents test data where each row is a test case and the columns are
+representing the inputs with the last column containing the expected result. Now
+let's say you want to split this data into two tensors, one representing the
+inputs and one the outputs.
+
+This might look like this:
+
+```csharp
+Tensor<float> testDataTable = GetTestData();
+
+// Take all rows and all but the last column
+Tensor<float> inputTable = testDataTable[.., ..-1];
+
+// Take all rows and only the last column
+Tensor<float> outputTable = testDataTable[.., -1..]; 
+```
+
+## Requirements
+
+### Goals
+
+* We need a type that both, plays well with existing concepts in the base class
+  library, as well as the intended syntax for C#
+* We need to ensure we can extend the notion of ranges in the base class library
+  in the future (for example, by supporting element types other than `Int32`).
+
+### Non-Goals
+
+* Supporting other numeric types or even offer a generic range type
+* Creating overloads on existing types that will accept `Range` (that doesn't
+  mean we shouldn't be doing it, but that it will be covered by a separate
+  design proposal).
+
+## Design
+
+### Proposed API
+
+```csharp
+namespace System {
+    public struct Range : IEnumerable, IEnumerable<int>, IEquatable<Range> {
+        public static Range Empty { get; }
+        public static Range All { get; }
+        public static Range Between(int startIndex, int endIndex);
+        public static Range From(int startIndex);
+        public static Range To(int endIndex);
+        public Range(int index, int length);
+        public int Index { get; }
+        public int EndIndex { get; }
+        public int Length { get; }
+        public bool IsUnbounded { get; }
+        public Range Bind(int length);
+        public override string ToString();
+        public override int GetHashCode();
+        public bool Equals(Range other);
+        public override bool Equals(object other);
```

Comment:
Nit: The parameter name for `override Equals(object)` should be `obj` to match the parameter name on [`object.Equals`](https://github.com/dotnet/corefx/blob/90027686dc6ee078ae190bfa471962efeab7a123/src/System.Runtime/ref/System.Runtime.cs#L1871). (Unlike `Equals(Range)` where it should be `other`).